tfjs-backend-webgpu/src/backend_webgpu.ts

```diff
@@ -827,49 +847,72 @@ export class WebGPUBackend extends KernelBackend {
       this.makeUniforms(programUniform)
     ];
 
-    const bindGroup = this.device.createBindGroup({
-      layout: pipeline.getBindGroupLayout(0),
-      entries: bindings.map((b, i) => ({binding: i, resource: b})),
-    });
-
-    this.ensureCommandEncoderReady();
-    const pass = this.getComputePass();
     const shouldTimeProgram = this.activeTimers != null;
-    if (shouldTimeProgram) {
-      if (this.supportTimeQuery) {
-        // tslint:disable-next-line:no-any
-        (pass as any).writeTimestamp(this.querySet, 0);
-      }
-    }
-    pass.setPipeline(pipeline);
-    pass.setBindGroup(0, bindGroup);
-    pass.dispatchWorkgroups(
-        program.dispatch[0], program.dispatch[1], program.dispatch[2]);
-    if (shouldTimeProgram) {
-      if (this.supportTimeQuery) {
-        // tslint:disable-next-line:no-any
-        (pass as any).writeTimestamp(this.querySet, 1);
+
+    // Currently only support parallel compilation for non-profiling mode.
+    const parallelCompilation =
+        env().getBool('ENGINE_COMPILE_ONLY') && !shouldTimeProgram;
+    if (!(program.shaderKey in this.pipelineCache)) {
+      if (parallelCompilation) {
+        this.pipelineCache[program.shaderKey] =
+            webgpu_program.compileProgramAsync(
+                this.device, program, inputsData, output);
+      } else {
+        this.pipelineCache[program.shaderKey] = webgpu_program.compileProgram(
+            this.device, program, inputsData, output);
       }
     }
-    this.dispatchNumberInEncoder++;
+    program.pipeline = this.pipelineCache[program.shaderKey];
 
     inputs.forEach(input => {
       this.commandQueueOwnedIds.add(input.dataId);
     });
     this.commandQueueOwnedIds.add(output.dataId);
 
+    if (!parallelCompilation) {
+      this.recordAndSubmit(program, bindings, shouldTimeProgram);
+      if (shouldTimeProgram) {
+        this.activeTimers.push({
+          name: program.constructor.name,
+          query: this.getQueryTime(this.querySet)
+        });
+      }
+    } else {
+      this.asyncProgramInfos.push({program, bindings});
+    }
+    return output;
+  }
+
+  private recordAndSubmit(
+      program: webgpu_program.WebGPUProgram, bindings: GPUBindingResource[],
+      shouldTimeProgram: boolean) {
+    program.bindGroup = this.device.createBindGroup({
+      layout: (program.pipeline as GPUComputePipeline).getBindGroupLayout(0),
```

Comment:
Here we should have to check that the `program.pipeline` is an object of `GPUComputePipeline` not `Promise<GPUComputePipeline>`.